File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-14.java
class Solution {
    public int characterReplacement(String s, int k) {
        Map<Character, Integer> count = new HashMap<>();
        int left = 0;
        int mostf = 0;
        int result = 0;

        for(int right = 0; right < s.length(); right++){
            char current = s.charAt(right);

            count.put(current, count.getOrDefault(current, 0) + 1);

            mostf = Math.max(mostf, count.get(current));


            while((right - left + 1) - mostf > k){
                char leftVal = s.charAt(left);
                count.put(leftVal, count.get(leftVal) - 1);
                left++;
            }


            result =  Math.max((right - left + 1), result);
        }

        return result;
    }
}

File: Data Structures & Algorithms/time-based-key-value-store/submission-14.py
class TimeMap:

    def __init__(self):
        self.key_value = defaultdict(list)

    def set(self, key: str, value: str, timestamp: int) -> None:
        self.key_value[key].append([value, timestamp])

    def get(self, key: str, timestamp: int) -> str:

        current = self.key_value.get(key, [])

        result = ""

        if len(current) == 0:
            return result
        
        left = 0
        right = len(current) - 1

        while(left <= right):

            middle = (left + right) // 2

            if current[middle][1] <= timestamp:
                result = current[middle][0]

            if current[middle][1] > timestamp:
                right = middle - 1
            else:
                left = middle + 1
            
        return result
            


        



        
